pallets/transaction-multi-payment/src/lib.rs

```diff
@@ -378,18 +379,33 @@ where
 	/// Note: The `fee` already includes the `tip`.
 	fn withdraw_fee(
 		who: &T::AccountId,
-		call: &T::RuntimeCall,
-		_info: &DispatchInfoOf<T::RuntimeCall>,
+		call: &<T as frame_system::Config>::RuntimeCall,
+		_info: &DispatchInfoOf<<T as frame_system::Config>::RuntimeCall>,
 		fee: Self::Balance,
 		_tip: Self::Balance,
 	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
 		if fee.is_zero() {
 			return Ok(None);
 		}
 
-		let currency = match call.is_sub_type() {
-			Some(Call::set_currency { currency }) => *currency,
-			_ => Pallet::<T>::account_currency(who),
+		let currency = if let Some(Call::set_currency { currency }) = call.is_sub_type() {
+			*currency
+		} else if let Some(pallet_utility::pallet::Call::<T>::batch { calls }) = call.is_sub_type() {
+			let first_call = &calls[0];
+			let currency = match first_call.is_sub_type() {
+				Some(Call::set_currency { currency }) => *currency,
+				_ => Pallet::<T>::account_currency(who),
+			};
+			currency
+		} else if let Some(pallet_utility::pallet::Call::<T>::batch_all { calls }) = call.is_sub_type() {
+			let first_call = &calls[0];
+			let currency = match first_call.is_sub_type() {
+				Some(Call::set_currency { currency }) => *currency,
+				_ => Pallet::<T>::account_currency(who),
+			};
+			currency
+		} else {
+			Pallet::<T>::account_currency(who)
 		};
```

Comment:
this can be simplified
```suggestion
let currency = if let Some(Call::set_currency { currency }) = call.is_sub_type() {
    *currency
} else {
    let calls = match call {
        Call::batch { calls } | Call::batch_all { calls } => calls,
        _ => return Pallet::<T>::account_currency(who),
    };
    let first_call = &calls[0];
    match first_call.is_sub_type() {
        Some(Call::set_currency { currency }) => *currency,
        _ => Pallet::<T>::account_currency(who),
    }
};
```

Reply:
you can even add force batch to it easily